uxarray/core/grid.py

```diff
@@ -283,6 +245,183 @@ def __from_ds__(self, dataset):
 
         dataset.close()
 
+    def __repr__(self):
+        """Constructs a string representation of the contents of a ``Grid``."""
+        prefix = "<uxarray.Grid>\n"
+        original_grid_str = f"Original Grid Type: {self.mesh_type}\n"
+        dims_heading = "Grid Dimensions:\n"
+        dims_str = ""
+        for key, value in zip(self._ds.dims.keys(), self._ds.dims.values()):
+            dims_str += f"  * {key}: {value}\n"
+
+        coord_heading = "Grid Coordinate Variables:\n"
+        coords_str = ""
+        if "Mesh2_node_x" in self._ds:
+            coords_str += f"  * Mesh2_node_x: {self.Mesh2_node_x.shape}\n"
+            coords_str += f"  * Mesh2_node_y: {self.Mesh2_node_y.shape}\n"
+        if "Mesh2_face_x" in self._ds:
+            coords_str += f"  * Mesh2_face_x: {self.Mesh2_face_x.shape}\n"
+            coords_str += f"  * Mesh2_face_y: {self.Mesh2_face_y.shape} \n"
+
+        connectivity_heading = "Grid Connectivity Variables:\n"
+        connectivity_str = ""
+        if "Mesh2_face_nodes" in self._ds:
+            connectivity_str += f"  * Mesh2_face_nodes: {self.Mesh2_face_nodes.shape}\n"
+        if "Mesh2_edge_nodes" in self._ds:
+            connectivity_str += f"  * Mesh2_edge_nodes: {self.Mesh2_edge_nodes.shape}\n"
+        if "Mesh2_face_edges" in self._ds:
+            connectivity_str += f"  * Mesh2_face_edges: {self.Mesh2_face_edges.shape}\n"
+        connectivity_str += f"  * nNodes_per_face: {self.nNodes_per_face.shape}\n"
+
+        return prefix + original_grid_str + dims_heading + dims_str + coord_heading + coords_str + \
+            connectivity_heading + connectivity_str
+
+    @property
+    def topology(self):
```

Comment:
We can, however we'd need to address how to handle accessing the attributes that are stored in `_ds`, especially for the case of MPAS meshes. I'd like to avoid having the user do `uxgrid._ds.attrs`.

Maybe adding a `uxgrid.attrs` or similar property?

Reply:
Yes, I, too, had thoughts about such options. I will give this a fresh thinking tomorrow and let you know if I can come up with some ideas

Reply:
Okay, I was able to form some thoughts. The term "attributes" is a bit open-ended (Xarray datasets have attributes, UGRID conventions attributes, etc.), so let's start with making sure we all are on the same page with that one first:

- We want our `Grid` object, i.e. `uxdataset.uxgrid` property, to represent the unstructured grid topology in the format UGRID convention suggests without needing additional structures. 
- Regarding the UGRID convention attributes, our `Grid` object has `Mesh2` "dummy variable" (as mentioned in the conventions) that already has the following required and "optionally required" attributes:
   
    <img width="367" alt="image" src="https://github.com/UXARRAY/uxarray/assets/32553057/ef24bcad-d3b6-4a4a-88da-2b401cbe8b78">

  plus the `long_name` string attribute
- I think `uxds.uxgrid.Mesh2` should be the go-to, sufficient topology attributes list when we regard to the UGRID convention attributes. That way, we can introduce 1:1 mapping between UXarray and UGRID conventions.
  - We are good with the required UGRID attributes already, but we can keep adding the [optional attributes](https://ugrid-conventions.github.io/ugrid-conventions/#2d-flexible-mesh-mixed-triangles-quadrilaterals-etc-topology:~:text=the%20arrays%20allow.-,The%20topology%20information%20is%20stored%20as%20attributes%20to%20a%20dummy%20variable%20(in%20the%20example%20below%20called%20%E2%80%9CMesh2%E2%80%9D)%20with%20cf_role%20mesh_topology.,-Required%20topology%20attributes) from the convention to our `Grid.Mesh2` over time. 

   That said, we should be good to go with the current structure without worrying about having an additional `Grid.topology`

- If you were referring to some MPAS-specific attributes instead, i.e. `uxds_mpas.uxgrid._ds.attrs`, with the following:

> We can, however we'd need to address how to handle accessing the attributes that are stored in _ds, especially for the case of MPAS meshes. I'd like to avoid having the user do uxgrid._ds.attrs.

  For instance, I can see the following attrs through this file "meshfiles/mpas/QU/mesh.QU.1920km.151026.nc":

   <img width="256" alt="image" src="https://github.com/UXARRAY/uxarray/assets/32553057/4e7db8af-270e-4b7e-a621-42dfbac20dae">

  Then, I think we may need to answer a few further questions:
   - Since we are representing any grid format in UGRID, would we really need/want to still keep track of format-specific attributes, or could we find a way to standardize them into UGRID format as well? 
      - If we decide to explicitly (without standardizing) keep track of such attributes, 
         - We could (and likely should) keep even such attributes under `Grid.Mesh2` maybe through some specification such as "format-specific attributes", 
         - or just like `Mesh2` stores the UGRID topology, we can consider introducing something like `Grid.original_attrs`

All in all, even if we cannot figure this format-specific attributes (e.g. MPAS), I still believe we are already good to go with `Grid.Mesh2` for the UGRID attributes.

Reply:
> The topology information is stored as attributes to a dummy variable (in the example below called “Mesh2”) with cf_role mesh_topology.

If I'm interpreting this correctly, it means that all attributes (which includes connectivity variables) are stored under `Mesh2` in the UGRID conventions. In our implementation, we are essentially treating our `Grid` object as `Mesh2`, storing both required and optional attributes (`node_coordinates`, `face_node_connectivity`, etc).

> our Grid object has Mesh2 "dummy variable" (as mentioned in the conventions) that already has the following required and "optionally required" attributes:

While this is correct, the attributes under `Mesh2` point to dictionary strings, not the actually variables. 
![image](https://github.com/UXARRAY/uxarray/assets/67855069/303a37d6-fb73-4f54-95cb-b0b24ffd6d30)

> Since we are representing any grid format in UGRID, would we really need/want to still keep track of format-specific attributes, or could we find a way to standardize them into UGRID format as well?

This would essentially require an extension of the UGRID conventions, since they don't appear to cover the cases that the MPAS attributes cover (sphere radius, periodicity, etc).

> We could (and likely should) keep even such attributes under Grid.Mesh2 maybe through some specification such as "format-specific attributes"

I like this idea. Only issue as I mentioned above is that `Mesh2` currently contains the dict of string attribute names of the connectivity variables, so adding parsed attributes into here could get cluttered.

> or just like Mesh2 stores the UGRID topology, we can consider introducing something like Grid.original_attrs

Going with something like `Grid.parsed_attrs` could be an appropriate place to store them. If it's alright with you, I think this solution would be the most clear

Reply:
>> The topology information is stored as attributes to a dummy variable (in the example below called “Mesh2”) with cf_role mesh_topology.

> If I'm interpreting this correctly, it means that all attributes (which includes connectivity variables) are stored under `Mesh2` in the UGRID conventions. 

Not really; i.e. not the variables themselves but only their names are stored under Mesh2.attributes so that the user can read the grid topology properly via using this as a map. For instance:

`Mesh2.edge_node_connectivity = "Mesh2_edge_nodes"`

> In our implementation, we are essentially treating our `Grid` object as `Mesh2`, storing both required and optional attributes (`node_coordinates`, `face_node_connectivity`, etc).

Not really, because the UGRID convention does not treat `Mesh2` as the whole grid topology either. It only says `Mesh` should keep attributes that maps to the actual variables of the grid topology, as I explained above. That said , I can tell both UGRID convention and our UXarray implementation treat `Grid` and `Mesh2` in the same way. 

Does that help?

Reply:
That clarification makes a lot more sense.

Reply:
> > Since we are representing any grid format in UGRID, would we really need/want to still keep track of format-specific attributes, or could we find a way to standardize them into UGRID format as well?
> 
> This would essentially require an extension of the UGRID conventions, since they don't appear to cover the cases that the MPAS attributes cover (sphere radius, periodicity, etc).

Great point, thanks! (Not directly related to this discussion but in general) This also implies representing MPAS in UGRID convention would diminish the analysis capabilities that'd come from the MPAS format.

Reply:
> > We could (and likely should) keep even such attributes under Grid.Mesh2 maybe through some specification such as "format-specific attributes"
> 
> I like this idea. Only issue as I mentioned above is that `Mesh2` currently contains the dict of string attribute names of the connectivity variables, so adding parsed attributes into here could get cluttered.
> 
> > or just like Mesh2 stores the UGRID topology, we can consider introducing something like Grid.original_attrs
> 
> Going with something like `Grid.parsed_attrs` could be an appropriate place to store them. If it's alright with you, I think this solution would be the most clear

I like this as well. Name sounds good! Yes, that's alright (though I may want to discuss options a bit further on this)! After you look into my additional explanations on `Mesh2` vs `Grid`, we can revisit this.

Reply:
> 



> That clarification makes a lot more sense.

Great, so that clarification is also what we should be attempting to document in our notebooks, usage examples, etc. 

That way, people who are familiar with the UGRD conventions should easily be able to get the fact that our `Grid.Mesh2.attributes` gives them the information about the grid topology, and using it, they could access Grid variables, e.g. `Mesh2_face_nodes`, etc.

Reply:
> > 
> 
> > That clarification makes a lot more sense.
> 
> Great, so that clarification is also what we should be attempting to document in our notebooks, usage examples, etc.
> 
> That way, people who are familiar with the UGRD conventions should easily be able to get the fact that our `Grid.Mesh2.attributes` gives them the information about the grid topology, and using it, they could access Grid variables, e.g. `Mesh2_face_nodes`, etc.

I think with the proper documentation, keeping the attributes under `Mesh2` would be good then!

Reply:
> I think with the proper documentation, keeping the attributes under `Mesh2` would be good then!

I agree!

Reply:
> > Going with something like `Grid.parsed_attrs` could be an appropriate place to store them. If it's alright with you, I think this solution would be the most clear
> 
> I like this as well. Name sounds good! Yes, that's alright (though I may want to discuss options a bit further on this)! After you look into my additional explanations on `Mesh2` vs `Grid`, we can revisit this.

Revisiting this:

- We have some `Grid` variables that could also be related to this discussion:
   `mesh_type:str`: The type of the parsed grid. 
   `source_grid:Path`: The grid source (file or URL) used to read this grid in
   `use_dual:bool`: Only meaningful for MPAS case. 

- We want something like `Grid.parsed_attrs` to keep track of the attributes of the parsed grid.

- Since all these four above are about the parsed grid, a few options:
   - We may want to group them into a single variable, maybe a dictionary of variables; or use the similar naming convention for each of them even if they are separate variables to ease understanding (e.g. all could start with "parsed_" etc.)
   - Or, we may want to keep `source_grid` and `mesh_type` as as but group `use_dual` (and potential future grid type-specific variables) and `parsed_attrs` as they are only meaningful for the parsed grid type.

- `grid_var_names` can be part of this decision and could go into one of above groups, namings, mentioned.. Or, at the least, we could consider making it a protected variable, just like `_ds`, and get rid of all related test cases that used it, because going forward, `Grid` variables with standard names should be sufficient and better for the sake of a unified UGRID naming.

Thoughts?

Reply:
> We want something like Grid.parsed_attrs to keep track of the attributes of the parsed grid.

In the most recent commit, this is handled by `Grid.Mesh2.attrs`, however if we would want a seperate `Grid.parsed_attrs` that can be done also. 

> Or, we may want to keep source_grid and mesh_type as as but group use_dual (and potential future grid type-specific variables) and parsed_attrs as they are only meaningful for the parsed grid type.

I didn't include these within the attributes since they aren't exactly "parsed" from the input dataset, we define them based on our own classification of the grid (`use_dual` is user defined). 

My definition of "parsed" is attributes that are taken from the input grid that don't have a direct 1:1 mapping to a UGRID attribute. In the case of MPAS, it's an attribute such as `on_a_sphere`. However for other model specifications, it could be anything from version number or anything else defined by that spec. 

I chose to include them as attributes of `Mesh2` since they are attributes that describe the mesh, but don't deserve their own dedicated property (such as `Grid.Mesh2_node_x`) 

> We may want to group them into a single variable, maybe a dictionary of variables; or use the similar naming convention for each of them even if they are separate variables to ease understanding (e.g. all could start with "parsed_" etc.)

A dictionary would be most appropriate for attributes.

> grid_var_names can also be part of this decision and could go into one of above groups, namings, mentioned..

I was planning on tackling this, but chose to save it for another PR since it's not of the highest priority right now, but I do agree.

Reply:
> > We want something like Grid.parsed_attrs to keep track of the attributes of the parsed grid.
> 
> In the most recent commit, this is handled by `Grid.Mesh2.attrs`, however if we would want a separate `Grid.parsed_attrs` that can be done also.

Per prior discussion, I think we shouldn't handle parsed attributes under `Mesh2` as it should keep the attributes for the UGRID format only. Instead, I'd suggest we do `Grid.parsed_attributes`.


> > Or, we may want to keep source_grid and mesh_type as as but group use_dual (and potential future grid type-specific variables) and parsed_attrs as they are only meaningful for the parsed grid type.
> 
> I didn't include these within the attributes since they aren't exactly "parsed" from the input dataset, we define them based on our own classification of the grid (`use_dual` is user defined).
>
> My definition of "parsed" is attributes that are taken from the input grid that don't have a direct 1:1 mapping to a UGRID attribute. In the case of MPAS, it's an attribute such as `on_a_sphere`. However for other model specifications, it could be anything from version number or anything else defined by that spec.

Makes sense; agreed!
 
> I chose to include them as attributes of `Mesh2` since they are attributes that describe the mesh, but don't deserve their own dedicated property (such as `Grid.Mesh2_node_x`)

Oh, I see, well, although this (them not deserving their own property) makes sense, I don't think listing them under `Mesh2` is a good idea (explained above). I believe `Grid.parsed_attributes` is still the best bet.


> > grid_var_names can also be part of this decision and could go into one of above groups, namings, mentioned..
> 
> I was planning on tackling this, but chose to save it for another PR since it's not of the highest priority right now, but I do agree.

Understood!

Reply:
Re: MPAS extra features and UGRID.  Our internal data model should use UGRID as a floor, not a ceiling.  If there's thing we can borrow from other mesh specs that make more functions possible, lets do that.

Reply:
> 



> Re: MPAS extra features and UGRID. Our internal data model should use UGRID as a floor, not a ceiling. If there's thing we can borrow from other mesh specs that make more functions possible, lets do that.

Good point; agree! Let's plan to do that as additional features, properties, etc. So, the design here for `Grid` should still make sense, and we can consider additional, model or function-specific attributes, functions in a case-by-case manner.

Reply:
> Understood

That's how I see it. Even with our current implementation, we have some additions that aren't described in the UGRID conventions (`nNodes_per_face`), so I think have this flexibility is powerful.

Reply:
> I believe Grid.parsed_attributes is still the best bet.

Working on getting this implementation in, thank you for the feedback!

Reply:
@erogluorhan 

One final consideration for this design. Should we utilize `grid._ds.attrs` to store these parsed attributes and provide a wrapper property to access them that does the following
```Python
@property
def parsed_attrs(self):
    return self._ds.attrs
```

Or instead should we have our own internal dictionary with the same property. 
```Python
@property
def parsed_attrs(self):
    return self._parsed_attrs
```

I lean towards `self._ds.attrs` since that's what our existing API already had, and considering our I/O modules work with respect to an **Input Datase**t and **Output Dataset**, not in terms of the `Grid` class itself, passing through the dictionary and property through to the modules would require a slight re-writing.

Reply:
> I lean towards `self._ds.attrs` since that's what our existing API already had, and considering our I/O modules work with respect to an **Input Datase**t and **Output Dataset**, not in terms of the `Grid` class itself, passing through the dictionary and property through to the modules would require a slight re-writing.

Agree, this is how I'd do it, too.